include/coap2/coap_session.h

```diff
@@ -490,4 +490,20 @@ coap_fixed_point_t coap_session_get_ack_random_factor(coap_session_t *session);
  */
 coap_tid_t coap_session_send_ping(coap_session_t *session);
 
+#define SESSIONS_ADD(e, obj) \
+  HASH_ADD(hh, (e), addr_info, sizeof((obj)->addr_info), (obj))
+
+#define SESSIONS_DELETE(e, obj) \
+  HASH_DELETE(hh, (e), (obj))
+
+#define SESSIONS_ITER(e, el, rtmp)  \
+  HASH_ITER(hh, (e), el, rtmp)
+
+#define SESSIONS_ITER_SAFE(e, el, rtmp) \
+for ((el) = (e); (el) && ((rtmp) = (el)->hh.next, 1); (el) = (rtmp))
+
+#define SESSIONS_FIND(e, k, res) {                     \
+    HASH_FIND(hh, (e), &(k), sizeof(k), (res)); \
```

Comment:
Not checked if this applies here but we should ensure that `coap_addr_tuple_t` objects are always memset to zero before filling with actual data. Otherwise, padding could lead to unexpected results when using the tuple as opaque hash key.  (For example, the Contiki version of `coap_address_t` has a `uint16_t` as last field.)

Reply:
I did think about this.  coap_make_session() memsets coap_session_t to 0, so should be OK here.
coap_read_session() copies session->addr_info to packet->addr_info before the packet is read into, along with the subsequent copy back of packet->addr_info to session->addr_info.

There is a possibility that there is another case that I have missed, but valgrind has not complained about uninitialized bytes - even with the contiki/lwip built versions of code and actual packet transfers between client / server.

Reply:
Thanks for the clarification, these functions seem to be save. There is at least one possible issue in `coap_read_session()` (see below).